core/src/main/java/hudson/model/Items.java

```diff
@@ -421,7 +421,7 @@ public static XmlFile getConfigFile(Item item) {
         return r;
     }
     private static <T extends Item> void getAllItems(final ItemGroup root, Class<T> type, List<T> r, Predicate<T> pred) {
-        List<Item> items = new ArrayList<>(((ItemGroup<?>) root).getItems());
+        List<Item> items = new ArrayList<>(((ItemGroup<?>) root).getItems(t -> t instanceof ItemGroup || (type.isInstance(t) && pred.test(type.cast(t)))));
         // because we add items depth first, we can use the quicker BY_NAME comparison
```

Comment:
Do we need all checks below after adding these predicates?

Reply:
Yes there is a sneaky edge case where type is `ItemGroup`. Now that I think about it I've forgotten to add a test for this and will be adding one.